### PR TITLE
chore: add a test harness for `<Dev/>`

### PR DESCRIPTION
**Pull request by threepointone**
This adds a basic test harness for the `<Dev/>` component. We'll add more tests and flesh out the harness a little more later, but figured we could land this and iterate on it.

### CODE REVIEW

**Comment by threepointone** [comment]
wip while I work through some issues I see with this

File: packages/wrangler/src/dev.tsx [hunk 1]
```diff
@@ -586,8 +579,9 @@ function useEsbuild({
       // esbuild already logs errors to stderr and we don't want to end the process
       // on build errors anyway so this is a no-op error handler
     });
-
-    return stopWatching;
+    return () => {
```
Comment (threepointone) [inline packages/wrangler/src/dev.tsx:582]: React ignores subsequent setups of the cleanup callback, so this needs to be a function on first run. I'm surprised there was no warning thrown here, I kinda remember us working on it back then. Might've been lost at some point.

File: packages/wrangler/src/module-collection.ts [hunk 2]
```diff
@@ -126,6 +126,12 @@ export default function createModuleCollector(props: {
             ),
           },
           async (args: esbuild.OnResolveArgs) => {
+            if (
```
Comment (threepointone) [inline packages/wrangler/src/module-collection.ts:129]: we specifically want to avoid entry points here

Comment (threepointone) [inline packages/wrangler/src/module-collection.ts:129]: we specifically want to avoid entry points here

File: packages/wrangler/src/dev.tsx [hunk 3]
```diff
@@ -448,35 +447,29 @@ function useLocalWorker(props: {
   return { inspectorUrl };
 }
 
+export interface DirectorySyncResult {
```
Comment (threepointone) [inline packages/wrangler/src/dev.tsx:450]: The types for `tmp.dirSync` call are wrong, so we redefined it here. I'll send a fix upstream soon.

Comment (threepointone) [inline packages/wrangler/src/dev.tsx:450]: The types for `tmp.dirSync` call are wrong, so we redefined it here. I'll send a fix upstream soon.

File: packages/wrangler/src/__tests__/dev2.test.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+              "textWrap": "wrap",
+            }
+          }
+        >
+          B to open a browser, D to open Devtools, L to turn off local mode, X to exit
+        </ink-text>
+      </ink-box>
+    `);
+    await TestRenderer.act(async () => {
+      // TODO: get rid of these sleep statements
```
Comment (threepointone) [inline packages/wrangler/src/__tests__/dev2.test.tsx:81]: The proper way to do this is to hook into the async process internally and listen when they shut down etc. That'll take some work, that I'll do soon.

Comment (threepointone) [inline packages/wrangler/src/__tests__/dev2.test.tsx:81]: The proper way to do this is to hook into the async process internally and listen when they shut down etc. That'll take some work, that I'll do soon.

**Comment by petebacondarwin** [comment]
Very excited to see this work in progress @threepointone. Great stuff!

**Comment by threepointone** [comment]
Let's land this and we can iterate on more stuff, don't want to keep rebasing as changes are being made to the rest of Dev elsewhere

File: packages/wrangler/src/dev/dev.tsx [hunk 5]
```diff
@@ -149,35 +148,29 @@ function Dev(props: DevProps): JSX.Element {
   );
 }
 
+export interface DirectorySyncResult {
```
Comment (threepointone) [inline packages/wrangler/src/dev/dev.tsx:151]: The types for `tmp.dirSync` call are wrong, so we redefined it here. I'll send a fix upstream soon.

File: packages/wrangler/src/dev/use-esbuild.ts [hunk 6]
```diff
@@ -84,7 +84,9 @@ export function useEsbuild({
       // on build errors anyway so this is a no-op error handler
     });
 
-    return stopWatching;
```
Comment (threepointone) [inline packages/wrangler/src/dev/use-esbuild.ts:87]: React ignores subsequent setups of the cleanup callback, so this needs to be a function on first run. I'm surprised there was no warning thrown here, I kinda remember us working on it back then. Might've been lost at some point.

File: packages/wrangler/src/dev/use-esbuild.ts [hunk 7]
```diff
@@ -84,7 +84,9 @@ export function useEsbuild({
...
+    return () => {
+      stopWatching && stopWatching();
```
Comment (petebacondarwin) [inline packages/wrangler/src/dev/use-esbuild.ts:88]: NIT:
```suggestion
      stopWatching?.();
```

File: packages/wrangler/src/__tests__/dev2.test.tsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        compatibilityDate={compatibilityDate}
+        compatibilityFlags={compatibilityFlags}
+        usageModel={usageModel}
+        bindings={bindings}
+        enableLocalPersistence={enableLocalPersistence}
+        zone={zone}
+      />
+    );
+  });
+  // @ts-expect-error How do I avoid the "used before assigned" error?
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/dev2.test.tsx:151]: I believe you just need to specify the type as undefinable...

```ts
let instance: TestRenderer.ReactTestRenderer|undefined;
```

File: packages/wrangler/package.json [hunk 9]
```diff
@@ -122,7 +124,8 @@
       "node_modules/(?!find-up|locate-path|p-locate|p-limit|yocto-queue|path-exists|execa|strip-final-newline|npm-run-path|path-key|onetime|mimic-fn|human-signals|is-stream|get-port)"
     ],
     "moduleNameMapper": {
-      "clipboardy": "<rootDir>/src/__tests__/helpers/clipboardy-mock.js"
+      "clipboardy": "<rootDir>/src/__tests__/helpers/clipboardy-mock.js",
+      "miniflare/cli": "<rootDir>/../../node_modules/miniflare/dist/src/cli.js"
```
Comment (petebacondarwin) [inline packages/wrangler/package.json:128]: Out of curiosity, why is this needed? Is it because of calls like:

```js
require.resolve("miniflare/cli")
```

**Review by petebacondarwin — APPROVED** [review]
Let's go go go

File: packages/wrangler/src/__tests__/dev2.test.tsx (hunk 8 above)
Comment (threepointone) [reply packages/wrangler/src/__tests__/dev2.test.tsx:151]: sweet, that works!

File: packages/wrangler/package.json (hunk 9 above)
Comment (threepointone) [reply packages/wrangler/package.json:128]: That's exactly it. `jest` hijacks the module loader so it can do stuff like mocks, but looks like it struggles with this specific call (because it's an es import? maybe, dunno.)
